### PR TITLE
(plugin)cloud::azure::policyinsights::policystates - new

### PR DESCRIPTION
**Pull request by sdepassio**

### CODE REVIEW

File: test/cloud/azure/custom/api.pl [hunk 1]
```diff
@@ -0,0 +1,113 @@
+use strict;
+use warnings;
+
+use Test2::V0;
+
+use FindBin qw( $RealBin );
+use lib "$FindBin::RealBin/../../../../src";
+use cloud::azure::custom::api;
```
Comment (garnier-quentin) [inline test/cloud/azure/custom/api.pl:8]: i don't understand that path for a test script

File: test/cloud/azure/custom/api.pl [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+#################
+
+my $urlPolicyStates = $api->azure_set_url(
+    providers => 'Microsoft.PolicyInsights/policyStates',
+    resource => 'default',
+    resource_group => 'myResourceGroup',
+    query_name => 'queryResults',
+    query => '$select=ResourceType, ResourceLocation, ResourceGroup, PolicyDefinitionName, ComplianceState'
+);
+is($urlPolicyStates, 'https://management.azure.com/subscriptions/00000000-0000-0000-0000-000000000001/resourceGroups/myResourceGroup/providers/Microsoft.PolicyInsights/policyStates/default/queryResults?api-version=2019-10-01&$select=ResourceType, ResourceLocation, ResourceGroup, PolicyDefinitionName, ComplianceState');
```
Comment (garnier-quentin) [inline test/cloud/azure/custom/api.pl:40]: I don't think it's very useful to test a string endpoint.

File: .gitignore [hunk 3]
```diff
@@ -0,0 +1 @@
+.idea/
```
Comment (garnier-quentin) [inline .gitignore:1]: please remove that file of the PR

Comment (garnier-quentin) [reply .gitignore:1]: Could you remove that file ?

Comment (sdepassio) [reply .gitignore:1]: this file is here to avoid inserting my IDE's files in the project

**Review by itoussies — APPROVED** [review]

**Review by itoussies — DISMISSED** [review]

**Review by garnier-quentin — APPROVED** [review]

**Comment by matoy** [comment]
Hi,

Are you sure this plugin works?
I could make it work.

image

**Comment by sdepassio** [comment]
I've fixed a problem here : https://github.com/centreon/centreon-plugins/tree/MON-18086-bug-on-azure-policies
If you could test it @matoy it would be great.

**Comment by garnier-quentin** [comment]
@sdepassio, you need to add:
```
query_form_post => '',
```

**Comment by matoy** [comment]
Hi,

just checked with the fix.
it's better but there are repetitive warnings:
Use of uninitialized value $resource_name in concatenation (.) or string at /usr/lib/centreon/plugins/centreon-plugins/cloud/azure/policyinsights/policystates/mode/compliance.pm line 107.
Use of uninitialized value $resource_name in concatenation (.) or string at /usr/lib/centreon/plugins/centreon-plugins/cloud/azure/policyinsights/policystates/mode/compliance.pm line 107.
Use of uninitialized value $resource_name in concatenation (.) or string at /usr/lib/centreon/plugins/centreon-plugins/cloud/azure/policyinsights/policystates/mode/compliance.pm line 107.
Use of uninitialized value $resource_name in concatenation (.) or string at /usr/lib/centreon/plugins/centreon-plugins/cloud/azure/policyinsights/policystates/mode/compliance.pm line 107.
Use of uninitialized value $resource_name in concatenation (.) or string at /usr/lib/centreon/plugins/centreon-plugins/cloud/azure/policyinsights/policystates/mode/compliance.pm line 107.
Use of uninitialized value $resource_name in concatenation (.) or string at /usr/lib/centreon/plugins/centreon-plugins/cloud/azure/policyinsights/policystates/mode/compliance.pm line 107.

i guess to make it more usable, it would be great to be able to filter.
can you add such option?
thanks.

**Comment by sdepassio** [comment]
You can filter by resourceLocation, resourceType and resourceGroup: are there other filters you can think of?

Regarding the warnings, I made another fix.

**Comment by sdepassio** [comment]
The loop and the new filter have been added.

**Comment by matoy** [comment]
thanks.

now i get:
UNKNOWN: Cannot decode response (add --debug option to display returned content)

even without using --policy-name parameter (same result when using it).

=> Send header: POST /subscriptions/xxx/providers/Microsoft.PolicyInsights/policyStates/default/queryResults?api-version=2019-10-01&%24select=ResourceId%2C+ResourceType%2C+ResourceLocation%2C+ResourceGroup%2C+PolicyDefinitionName%2C+IsCompliant%2C+ComplianceState HTTP/1.1
Host: management.azure.com
Accept:application/json
Authorization:Bearer xxx
Content-Length: -1
Content-Type: application/x-www-form-urlencoded
Expect: 100-continue

=> Recv header: HTTP/1.1 400 Bad Request
=> Recv header: Content-Type: text/html; charset=us-ascii
=> Recv header: Server: Microsoft-HTTPAPI/2.0
=> Recv header: Date: Thu, 30 Mar 2023 09:12:59 GMT
=> Recv header: Connection: close
=> Recv header: Content-Length: 374
=> Recv header:
=> Recv data: <!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01//EN""http://www.w3.org/TR/html4/strict.dtd">
<HTML><HEAD><TITLE>Bad Request</TITLE>
<META HTTP-EQUIV="Content-Type" Content="text/html; charset=us-ascii"></HEAD>
<BODY><h2>Bad Request - Invalid Content Length</h2>

**Pull request merged**